source/common/common/logger.h

```diff
@@ -417,33 +419,40 @@ template <Id id> class Loggable {
   } while (0)
 
 #define ENVOY_LOG_EVERY_POW_2(LEVEL, ...)                                                          \
   do {                                                                                             \
-    static auto* count = new std::atomic<uint64_t>();                                              \
-    if (std::bitset<64>(1 + count->fetch_add(1)).count() == 1) {                                   \
-      ENVOY_LOG(LEVEL, ##__VA_ARGS__);                                                             \
+    if (ENVOY_LOG_COMP_LEVEL(ENVOY_LOGGER(), LEVEL)) {                                             \
+      static auto* count = new std::atomic<uint64_t>();                                            \
+      if (std::bitset<64>(1 + count->fetch_add(1)).count() == 1) {                                 \
+        ENVOY_LOG(LEVEL, ##__VA_ARGS__);                                                           \
+      }                                                                                            \
     }                                                                                              \
   } while (0)
 
 // This is to get us to pass the format check. We reference a real-world time source here.
-// I think it's not easy to mock/simulate time here as it stands today anyway.
+// We'd have to introduce a singleton for a time source here, and consensus was that avoiding
+// that is preferable.
 using t_logclock = std::chrono::steady_clock; // NOLINT
 
 #define ENVOY_LOG_PERIODIC(LEVEL, CHRONO_DURATION, ...)                                            \
   do {                                                                                             \
-    static auto* last_hit = new std::atomic<uint64_t>();                                           \
-    auto last = last_hit->load();                                                                  \
-    const auto now = t_logclock::now().time_since_epoch().count();                                 \
-    if ((now - last) >                                                                             \
-            std::chrono::duration_cast<std::chrono::nanoseconds>(CHRONO_DURATION).count() &&       \
-        last_hit->compare_exchange_strong(last, now)) {                                            \
-      ENVOY_LOG(LEVEL, ##__VA_ARGS__);                                                             \
+    if (ENVOY_LOG_COMP_LEVEL(ENVOY_LOGGER(), LEVEL)) {                                             \
+      static auto* last_hit = new std::atomic<uint64_t>();                                         \
+      auto last = last_hit->load();                                                                \
+      const auto now = t_logclock::now().time_since_epoch().count();                               \
+      if ((now - last) >                                                                           \
```

Comment:
Change `last_hit` to a type based on `int64_t`. The standard c++ `time_since_epoch().count()` returns `int64_t` value (`decltype(now)` is ` const int64_t`) and there should be a warning about type conversion.

Reply:
nice catch, thanks